### ISSUE TITLE
Better support when multiple renderers are installed

### ISSUE THREAD

**Issue by misolori** [microsoft/vscode#125876]
Steps to Reproduce:

1. Install the Jupyter extension
2. Install another notebook extension w/ a custom renderer ([httpBook](https://marketplace.visualstudio.com/items?itemName=anweber.httpbook))
3. Open a file from custom renderer (.http), everything looks fine
    <img width="670" alt="CleanShot 2021-06-09 at 13 56 27@2x" src="https://user-images.githubusercontent.com/35271042/121428314-814c4200-c92a-11eb-804e-61de73732017.png">
4. Open a Jupyter notebook (.ipynb)
5. :bug: all outputs have defaulted to the custom renderer
    <img width="891" alt="CleanShot 2021-06-09 at 13 58 10@2x" src="https://user-images.githubusercontent.com/35271042/121428501-bc4e7580-c92a-11eb-8910-db9c03dc9fd5.png">
    <img width="647" alt="CleanShot 2021-06-09 at 13 58 19@2x" src="https://user-images.githubusercontent.com/35271042/121428530-c53f4700-c92a-11eb-8689-a95e62b30465.png">

I would expect that a custom renderer that only works with certain filetypes (http) to only appear for those notebooks, not all notebooks. I'd also hope for an easy way to set the default renderer for certain notebook types.

**Comment by rebornix** [comment]
Not sure yet if this should be a candidate, can't recall when/why the user setting to specify which extension/renderer has higher priority is gone.

**Comment by connor4312** [comment]
This is actually working 'as designed'. The mimetype for httpbook's [custom renderer](https://github.com/AnWeber/httpbook/blob/61240a4187eb5cd29edcf5590fb5b26f4b9e1c74/package.json#L72-L85) is text/html. When the user selects that, we remember that they prefer that as their renderer for the given mimetype.

However maybe we should only remember that in the context of a single `notebookType` rather than globally.

**Comment by rebornix** [comment]
The catch here is users are very likely not aware that this extension contributes `text/html` renderer and might think this is a bug. It's not visible anywhere in the notebook editor that you can switch mimetypes or renderers.

Also I'm wondering if some of the basic mime type renderers in the core should have higher priority than the custom ones. For example `text/html` is trivial and that's what users should get oob no matter if there is a renderer for `text/html`. Otherwise every extension will implement their own `text/html` renderer if they want the outputs rendering to be consistent for basic mime types.

> When the user selects that, we remember that they prefer that as their renderer for the given mimetype

@connor4312 this is not working right now.

**Comment by connor4312** [comment]
I think that's a good idea _if_ we also have some notion of extension affinity, where a custom renderer for a built-in mimetype is preferred if the output was created by the same extension who contributed the renderer. If this is not the case, then it's a poor experience for notebooks that intend to have a custom renderer,  since their users will manually need to switch their renderer.

Or we just disallow people overwriting built-in mimetypes, and say to use a custom mimetype instead.

**Comment by rebornix** [comment]
Thought about this again, as a renderer author, my renderer offers rich rendering for `application/vnd.code.notebook.error`, which is also builtin. I personally would expect the custom renderer to be the default renderer for `application/vnd.code.notebook.error` in all types of notebooks (if it's the only custom one).

Considering how I built the renderer, I don't think we can disallow people overwriting builtin mimetypes.

**Comment by tanhakabir** [comment]
Yeah I also don't like disallowing overriding builtin mimetypes but we can say that the default builtin mimetype is the preferred one unless the user selects a different one for the viewtype or the renderer extension itself contributed the renderer

**Comment by DonJayamanne** [comment]
> one unless the user selects a different one for the viewtype or the renderer extension itself contributed the renderer

What happens if I select XYZ renderer for rendering HTML when looking at Notebook ABC (where both were contributed by two seprate extensions). Does this mean that XYZ will be used to render HTML output in httpbook?

I think this is where the challenge is, as the user may not expect that to happen.

**Comment by connor4312** [comment]
> What happens if I select XYZ renderer for rendering HTML when looking at Notebook ABC (where both were contributed by two seprate extensions). Does this mean that XYZ will be used to render HTML output in httpbook?

This would be in addition to the `notebookType ` association, so it would only set it in ABC.

**Comment by DonJayamanne** [comment]
So, something like a generic renderer for `application/vnd.code.notebook.error` would only apply for a specific notebook type. is that correct?

**Comment by tanhakabir** [comment]
@nat1craft mentioned in REST Book that the last output renderer he selected isn't the default in the next cell run. @connor4312 is this expected?

**Comment by AnWeber** [comment]
@tanhakabir I can confirm that the selected editor is not the default for the next run. 

I saw the separation between Notebook and NotebookRenderer and the possibility of different renderers for the same Mimetype as a feature. One use case for this would be rendering application/geo+json. Partly I want to see the representation as JSON, partly I want to view the data in a map view. However, I would leave it up to the user to decide how the default display should be (https://github.com/microsoft/vscode/issues/123884#issuecomment-843319841).

**Comment by RandomFractals** [comment]
yep. you should try this too: https://github.com/RandomFractals/vscode-data-table/issues/23#issuecomment-874833269

![data-table-chicago-traffic](https://user-images.githubusercontent.com/656833/124622869-17e42400-de41-11eb-9119-bce89899fa1d.png)

**Comment by RandomFractals** [comment]
this might need further investigation on your end for CSV data loading and renderers with support for mulitple mime types, csv and json in my prototype ... https://github.com/tanhakabir/rest-book/issues/114

**Comment by connor4312** [comment]
How about this algorithm. For any given output with mimetype X:

- If the user has explicitly selected a renderer for X, in this notebook type and workspace, use that.
- Otherwise, if the extension providing the notebook also provides a renderer for X, use that.
- Otherwise, if there is a built-in renderer for X, use that. (From previous comments)
- Otherwise, if there is another extension who renders mimetype X, use that.

Whenever a user changes the output renderer, persist the preference to use that renderer for mimetype X within the current notebook type in workspace storage. (Effectively `Map<NotebookType, Map<Mimetype, RendererId>>`)
